src/os/bluestore/BlueStore.cc

```diff
@@ -16535,3 +16698,1473 @@ void RocksDBBlueFSVolumeSelector::dump(ostream& sout) {
+    bufferlist              header_bl;
+    //dout(5) << " header = \n" << header << dendl;
+    encode(header, header_bl);
+    crc = header_bl.crc32c(crc);
+    encode(crc, header_bl);
+    p_handle->append(header_bl); 
+  }
+
+  crc = -1;					 // reset crc
+  extent_t        buffer[MAX_EXTENTS_IN_BUFFER]; // 64KB
+  extent_t       *p_curr          = buffer;
+  const extent_t *p_end           = buffer + MAX_EXTENTS_IN_BUFFER;
+  uint64_t        extent_count    = 0;
+  uint64_t        allocation_size = 0;
+  auto iterated_allocation = [&](uint64_t extent_offset, uint64_t extent_length) {
+    if (extent_length == 0) {
+      derr <<  __func__ << "" << extent_count << "::[" << extent_offset << "," << extent_length << "]" << dendl;
+      ret = -1;
+      return;
+    }
+    //dout(5) <<  "" << extent_count << "[" << extent_offset << "," << extent_length << "]" << dendl;
+    p_curr->offset = HTOCEPH_64(extent_offset);
+    p_curr->length = HTOCEPH_64(extent_length);
+    extent_count++;
+    allocation_size += extent_length;
+    p_curr++;
+
+    if (p_curr == p_end) {
+      crc = flush_extent_buffer_with_crc(p_handle.get(), (const char*)buffer, (const char*)p_curr, crc);
+      //dout(5) <<  " extent_count=" << extent_count << ", crc=" << crc << dendl;
+      p_curr = buffer; // recycle the buffer
+    }
+  };
+  allocator->dump(iterated_allocation);
+  // if got null extent -> fail the operation
+  if (ret != 0) {
+    derr << "Illegal extent, fail store operation" << dendl;
+    derr << "invalidate using bluefs->truncate(p_handle, 0)" << dendl;
+    bluefs->truncate(p_handle.get(), 0);
+    return -1;
+  }
+
+  // if we got any leftovers -> add crc and append to file
+  if (p_curr > buffer) {
+    crc = flush_extent_buffer_with_crc(p_handle.get(), (const char*)buffer, (const char*)p_curr, crc);
+    //dout(5) <<  " extent_count=" << extent_count << ", crc=" << crc << dendl;
+  }
+
+  {
+    allocator_image_trailer trailer(timestamp, s_format_version, s_serial, extent_count, allocation_size);
+    bufferlist trailer_bl;
+    //dout(5) << "trailer=\n" << trailer << dendl;
+    encode(trailer, trailer_bl);
+    uint32_t crc = -1;
+    crc = trailer_bl.crc32c(crc);
+    encode(crc, trailer_bl);
+    p_handle->append(trailer_bl);
+  }
+  
+  bluefs->fsync(p_handle.get());  
+  bluefs->truncate(p_handle.get(), p_handle->pos);
+
+  utime_t duration = ceph_clock_now() - start_time;
+  dout(5) <<"WRITE-extent_count=" << extent_count << ", file_size=" << p_handle->file->fnode.size << dendl;
+  dout(5) <<"p_handle->pos=" << p_handle->pos << " WRITE-duration=" << duration << " seconds" << dendl;
+
+  return 0;
+}
+
+//-----------------------------------------------------------------------------------
+Allocator* BlueStore::create_allocator(uint64_t bdev_size, string type) {
+  dout(5) << "type=" << type << dendl;
+  // create allocator
+  uint64_t alloc_size = min_alloc_size;
+#ifdef HAVE_LIBZBD
```

Comment:
This is used to temporary store allocations in-memory. Zone - specific allocations policies do not apply.